vpr/src/place/cut_spreader.cpp

```diff
@@ -0,0 +1,816 @@
+#include "cut_spreader.h"
+#include <iostream>
+#include <vector>
+#include <queue>
+#include <cstdlib>
+
+#include "analytic_placer.h"
+#include "vpr_types.h"
+#include "vtr_time.h"
+#include "globals.h"
+#include "vtr_log.h"
+
+CutSpreader::CutSpreader(AnalyticPlacer* analytic_placer, VprContext& vpr_ctx, t_logical_block_type_ptr cell_t)
+    : ap(analytic_placer)
+    , device_ctx(vpr_ctx.device())
+    , clb_nlist(vpr_ctx.clustering().clb_nlist)
+    , place_ctx(vpr_ctx.mutable_placement())
+    , blk_type(cell_t) {
+    // get "fast_tile" for blk_type, quick look up by x, y coords
+    ft.resize(ap->max_x, std::vector<std::vector<t_pl_loc>>(ap->max_y));
+    for (auto& tile : blk_type->equivalent_tiles) {
+        for (auto sub_tile : tile->sub_tiles) {
+            auto result = std::find(sub_tile.equivalent_sites.begin(), sub_tile.equivalent_sites.end(), blk_type);
+            if (result != sub_tile.equivalent_sites.end()) {
+                for (auto loc : ap->legal_pos.at(tile->index).at(sub_tile.index)) {
+                    ft.at(loc.x).at(loc.y).emplace_back(loc);
+                }
+            }
+        }
+    }
+}
+
+void CutSpreader::run() {
```

Comment:
This needs a big comment.  What is the expected input? (Placement state, but where is it stored and what state is it expected to be in? ).  Where is the output stored and is it fully legal or just more spread out?  What is the algorithm?  What is the work queue?

Reply:
Also looks like regions had to previously be set up.  Who set them up and how big are they?
Some of these comments may be file scope (how does the whole cut spreader work)?